cilium/CFP-14339-cilium-mtu-model.md

```diff
@@ -0,0 +1,292 @@
+﻿# CFP-14339: Cilium MTU model
+
+**SIG:** SIG-Datapath
+
+**State:** Dormant[^a]
+
+**Begin Design Discussion:** Feb 26, 2021
+
+**Authors:** Joe Stringer <joe@cilium.io>
+
+**Issue:** https://github.com/cilium/cilium/issues/14339
+
+## Summary
+
+Cilium 1.10 or earlier does not provide deterministic Ethernet MTU (Maximum
+Transmission Unit) size configuration in environments where multiple networks
+are exposed to the nodes where Cilium runs. This feature aims to configure MTU
+consistently and optimally based on the available networks.
+
+### Motivation
+
+Cilium issue #14339[^o], Cilium is using wrong interface for MTU detection.
+
+It is common in cloud environments to have multiple networks exposed to Cilium
+where the MTU of the interfaces attached to those networks may vary quite
+widely, such as allowing 1500 byte ethernet payloads via one network while
+supporting as much as 9000 byte payloads via another network. Misconfiguration
+of the MTU across different nodes can cause connectivity disruption,
+particularly for UDP packets.
+
+In some user environments, the Cilium 1.9 MTU calculation behaviour is
+inconsistent as it is determined by either:
+
+* `--mtu` flag (administrator override)
+* `InitDefaultPrefix()` configures the node IP address
+   * `->firstGlobalV4Addr()`
+      * List all IPs from interfaces, except for interfaces that are down and
+        docker interfaces
+      * Pick the first one in the sorted list
+
+
+We also currently expose only MTU and not MRU. Rather than considering transmit
+and receive separately to determine the maximum sizes that can be viably sent
+vs. received depending on the attached networks, we select a single MTU and
+derive the MRU based on the configured/detected MTU and other configuration
+such as ipsec, tunnel. This means that if the MTU we detected is incorrect,
+then we may configure the MRU in a way that may cause traffic drops on receive
+in cases where the packets could otherwise have been successfully received.
```

Comment:
I think that the following text adjustment would more accurately use the term MRU:

```suggestion
Cilium exposes only MTU and not MRU for all paths. Internally the code is not structured to clearly differentiate the maximum sizes that can be viably sent
vs. received depending on the attached networks. The logic selects a single MTU, treats
this as the MRU for all paths, then derives the MTU for tunnel paths based on other configuration
such as ipsec, tunnel. This means that if the MTU we detected is incorrect,
then we may configure either (a) device MTU on receive side in a way that may cause traffic drops, or (b) route MTU on transmit side in a way that may transmit packets which cannot be delivered.
```